### ISSUE TITLE
[2014 CVPR Oral] Rich feature hierarchies for accurate object detection and semantic segmentation

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#28]
![image](https://user-images.githubusercontent.com/49643709/123216290-5524e980-d504-11eb-9e33-0052d6ec3bea.png)

R-CNN 논문으로 처음으로 detection에서 CNN을 이용한 논문

Object detecion을 다음과 같이 2가지 방법으로 구분하여 task 수행 (2-shot)
1. Region Proposals
2. Feature Extraction and Prediction

Train, inference, experiments에 대하여 간단하게 설명

## 1. Train
![image](https://user-images.githubusercontent.com/49643709/123219159-8a7f0680-d507-11eb-9152-2366d188281d.png)

- Feature extraction
    - Region proposal 중 ground-truth box와 IoU가 0.5 이상인 경우에만 positive, 그 외에는 negative
    - 32 positive + 96 negative 총 128 batch
    - Fine-tuning을 통해 CNN 모델 학습 (SGD with 0.001 learning rate, pre-trained on ILSVRC-2012)
    - CNN 모델은 AlexNet을 주로 사용
- Object classifier
    - 1개의 class마다 1개의 linear SVM model 사용
    - Region proposal 중 ground-truth box와 IoU가 0.5 이상인 경우에만 positive, IoU가 0.3 이하인 경우는 negative (그 사이는 무시)
    - Standard hard negative mining method를 이용하여 linear SVM 학습
    - Linear SVM를 사용하지 않고 CNN에서 fc layer를 추가하여 softmax를 통해 prediction을 하면 성능이 저하됨
- Bounding box regressor
    - Bounding box regressor 없이 성능 평가를 하면 localization error로 인해 성능 저하가 큼
    - Pool5 feature와 그에 해당하는 region이 주어졌을 때, linear regression을 통해 bounding box를 수정

## 2. Inference

1. Fast mode selective search를 통해 2000개의 region proposal
2. 각 region proposal에서 16 pixels만큼 더 크게 bounding box를 잡아서 fixed 227 * 227 pixel size로 warp
3. Warp image를 CNN을 통해 4096 feature vector로 extraction
4. Extracted feature vector를 각 class에 대해 학습된 SVM을 이용하여 score을 매김
5. 1개의 image에 대하여 2000개의 region proposal에 대한 score을 얻음
6. Thresholding, greedy NMS를 이용하여 bounding box prediction

### 3.1. Visualization
![image](https://user-images.githubusercontent.com/49643709/123216630-b947ad80-d504-11eb-8462-4b399b69c229.png)

CNN이 학습한 것이 무엇인지 알아보기 위한 실험 수행

Feature vector unit activation의 크기를 기준으로 proposal을 정렬한 후, NMS를 통해 proposal을 줄임

Proposal을 보면서 저자들은 CNN 모델이 small number of class-tuned features와 더불어 shape, texture, color, material properties와 같은 정보를 종합하여 학습한다고 분석함

### 3.2. Ablation Studies
![image](https://user-images.githubusercontent.com/49643709/123218491-bf3e8e00-d506-11eb-8917-835c37d8b2fd.png)
![image](https://user-images.githubusercontent.com/49643709/123220424-e8f8b480-d508-11eb-956c-8db205330c1c.png)

- Performance layer-by-layer, without fine-tuning
    - CNN 모델을 fine-tuning하지 않고 ILSVRC-2012 pre-trained model로만 performance measure
    - Pool5, fc6, fc7의 output간의 결과를 비교
    - Pool5만으로도 충분히 괜찮은 결과를 얻을 수 있음 (fc layer을 제거하면 모델이 매우 가벼워짐)
    - 이를 통해 CNN의 representational power는 convolution layer로부터 나온다는 것을 알 수 있음
- Performance layer-by-layer, with fine-tuning
    - Fine-tuning을 하면 성능이 확실히 좋아짐
    - Pool5 결과와 fine-tuning하지 않은 pool5 결과를 비교하면 성능이 비슷함
    - 이를 통해 ImageNet으로 배운 feature가 general함을 확인할 수 있음
    - 성능 향상은 domain-specific한 non-linear classifier가 학습하여 올랐다고 볼 수 있음
- CNN model comparison
    - AlexNet과 VGG 16 모델을 이용하여 CNN model의 차이가 성능에 어떤 영향을 미치는지 확인
    - VGG 16 모델을 사용하면 성능이 확실히 더 올라가지만, 모델이 무거워 시간이 오래 걸린다는 단점이 있음
- Warp
    - Region proposal을 통해 image patch를 얻게 되면 CNN 모델에 넣어주기 위해 size를 맞춰주어야함
    - Context padding을 넣어준 후 227 * 227 pixel size로 맞춰주는 것이 가장 성능이 좋았음

**Comment by Jasonlee1995** [comment]
![image](https://user-images.githubusercontent.com/49643709/123226963-15173400-d50f-11eb-9db8-ea92ee66f370.png)
![image](https://user-images.githubusercontent.com/49643709/123226986-1b0d1500-d50f-11eb-818c-5f9f131cfc03.png)

- d_x, d_y, d_w, d_h 함수 : linear function which input is pool5 feature
- 즉 pool5를 통해 region proposal을 어느정도 수정한다는 의미
- 실험에서 lambda = 1000으로 세팅하였으며, regularization을 잘 세팅하는 것이 중요하다고 함
- 또한 (P, G)를 잘 세팅해야 학습이 됨 (region proposal P가 ground truth G와 거리가 멀다면 말이 안되기에)
- 이를 위해 nearness가 될 수 기준을 IoU로 설정 (IoU가 0.6 이상인 경우 학습 데이터 pair로 인정)
- 참고로 bounding box regressor는 class-specific함 (즉 1개의 class마다 1개의 bounding box regressor가 존재)
